runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,417 @@
+local util = require('vim.lsp.util')
+local log = require('vim.lsp.log')
+local ms = require('vim.lsp.protocol').Methods
+local api = vim.api
+
+local M = {}
+
+---@class (private) vim.lsp.folding_range.BufState
+---
+---@field version? integer
+---
+--- Never use this directly, `renew()` the cached foldinfo
+--- then use on demand via `row_*` fields.
+---
+--- Index In the form of client_id -> ranges
+---@field client_ranges table<integer, lsp.FoldingRange[]?>
+---
+--- Index in the form of row -> foldlevel
+---@field row_level table<integer, [integer, ">" | "<"?]?>
+---
+--- Index in the form of start_row -> kinds
+---@field row_kinds table<integer, table<lsp.FoldingRangeKind, true?>?>>
+---
+--- Index in the form of start_row -> collapsed_text
+---@field row_text table<integer, string?>
+
+---@type table<integer, vim.lsp.folding_range.BufState?>
+local bufstates = {}
```

Comment:
Just for reference: this is yet another case of a pattern I think we discussed generalizing, right? (can't find the PR/issue ...)

Reply:
Yes, there was [a draft PR](https://github.com/neovim/neovim/pull/30841) before. Many of lewis's earlier reviews focused on ensuring that these bufstates can be correctly destroyed to avoid potential memory leaks.

I still haven't given up on this idea, so I kept the code related to managing bufstates in a single `setup` function there so that it can truly generalize to other modules in the future. Currently, it should ensure that nothing remains on the unloaded buffer. It would be better if we could delete it when there is no corresponding method supported on the buffer, as lewis suggested. Still, for now, the actual effect is not very significant, and the risk is relatively high.